### DISCUSSION TITLE
How to upload duplicate image in different ocg layer

### DISCUSSION THREAD

**Discussion by navjotmakkar** [pymupdf/PyMuPDF#1857] [Looking for help] [answered]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
I have 3 pdf, among with 2 of them has same logo. I am processing each pdf individually where I am extracting the logo from them and adding them in different ocg layer created using add_ocg() and image is added using insert_image(). 
Since the two images(logo) are same, when I am adding it to the second layer it takes a reference from already added one, instead of handling it separately. Due to this if I disable the first layer(layer name- CW1.pdf) in adobe reader, the logo of another layer(layer name- WW1.pdf) also get disabled with it.
Also, we tried changing value of xref parameter of insert_image() to None, negative, random integer but the issue persists.

## To Reproduce (mandatory)
Add same image in two different ocg layer.

## Expected behavior (optional)
We expected that the image added in seperate ocg layers should be handled within that layer only, instead of taking reference from already added one. The last logo of 'Thames Water' in uploaded pdf should be in sync with the 'WW1.pdf' ocg layer.

## Screenshots (optional)
Please find attached the final processed pdf created using pymupdf. 
[layered_output.pdf](https://github.com/pymupdf/PyMuPDF/files/9259583/layered_output.pdf)

## Your configuration (mandatory)
 - Operating system, potentially version and bitness
 - Python version, bitness
 - PyMuPDF version, installation method (**wheel** or **generated** from source).

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

**Comment by JorjMcKie** [comment]
This is probably not a bug, but an unimplemented feature.
When adding an image the insertion method will automatically check (on the binary image content level), whether it already exists.
If so, it will not be added one more time, but just a reference to it will be inserted in the page's object definition. The reasons for this approach are obvious presumably.
This implies (and this is documented) that any optional content specification must be ignored in this case. You can only modify the image's original object defintion and give this a desired `oc=nnn` number.
In other words: the same image cannot exist twice in the same PDF with different OC references.

**Comment by JorjMcKie** [comment]
Your problem should go away if you are able to make the second insertion with an image version that has a different binary content.
E.g. if you extracted a JPG originally, then make a pixmap of it and use this one in the image insertion method. This will lead to inserting a PNG version of the logo.

**Comment by navjotmakkar** [comment]
> Your problem should go away if you are able to make the second insertion with an image version that has a different binary content. E.g. if you extracted a JPG originally, then make a pixmap of it and use this one in the image insertion method. This will lead to inserting a PNG version of the logo.

I am processing multiple(three) pdf in a for loop, I tried creating pixmap instead of using direct filename and since 1st and 3rd logo are same I guess there pixmap are also same as it is still taking the reference from the first one instead adding new image. pixmap created using thsi line- pix = fitz.Pixmap(logo_crop_filename)
Any suggestions how to create different version of same image without modifying it appearance? Also, currently its just three 3 pdf but there can be more so everytime there should be a different version of image on binary level but still look same.

**Comment by JorjMcKie** [comment]
I have converted the issue to a discussion. Let's continue dealing with the situation here.

I was just thinking about tricking out my logic which is responsible for preventing duplicate images.
Whenever I insert an image I store away its xref in `doc.InsertedImages`.
This is a dictionary with a key equal to the MD5 code of the binary image content, and the value being the inserted xref.
What you could try is locating the key/value pair in that dictionary and remove the item after an `insert_image` whenever you want to insert the image again:
Method `insert_image` returns the xref of the insertion. Lookup that value in `doc.InsertedImage`: `item = [(k,v) for k,v in doc.InsertedImages.items() if v==xref][0]`.
This list should have exactly one item, so you can do `del doc.InsertedImages[item[0]]`.
After that you should be able to fully insert same image again.

**Comment by JorjMcKie** [comment, marked as answer]
If you want to be lazy, you could of course also do `doc.InsertedImages = {}` after each `insert_image`.
